lib/puppet/functions/gitlab_ci_runner/register_to_file.rb

```diff
@@ -0,0 +1,57 @@
+require_relative '../../../puppet_x/gitlab/runner.rb'
+require 'fileutils'
+
+# A function that registers a Gitlab runner on a Gitlab instance, if it doesn't already exist,
+# _and_ saves the retrived authentication token to a file. This is helpful for Deferred functions.
+Puppet::Functions.create_function(:'gitlab_ci_runner::register_to_file') do
+  # @param url The url to your Gitlab instance. Please only provide the host part (e.g https://gitlab.com)
+  # @param regtoken Registration token.
+  # @param runner_name The name of the runner. Use as identifier for the retrived auth token.
+  # @param filename The filename where the token should be saved.
+  # @param additional_options A hash with all additional configuration options for that runner
+  # @return [String] Returns the authentication token
+  # @example Using it as a Deferred function
+  #   gitlab_ci_runner::runner { 'testrunner':
+  #     config               => {
+  #       'url'              => 'https://gitlab.com',
+  #       'token'            => Deferred('gitlab_ci_runner::register_runner_to_file', [$config['url'], $config['registration-token'], 'testrunner'])
+  #       'executor'         => 'shell',
+  #     },
+  #   }
+  #
+  dispatch :register_to_file do
+    # We use only core data types because others aren't synced to the agent.
+    param 'String[1]', :url
+    param 'String[1]', :regtoken
+    param 'String[1]', :runner_name
+    optional_param 'Hash', :additional_options
+    optional_param 'String[1]', :filename
+    return_type 'String[1]'
+  end
+
+  def register_to_file(url, regtoken, runner_name, additional_options = {}, filename = "/etc/gitlab-runner/auth-token-#{runner_name}")
+    if File.exist?(filename)
+      authtoken = File.read(filename).strip
+    else
+      return 'DUMMY-NOOP-TOKEN' if closure_scope['facts'].key?('clientnoop')
```

Comment:
I think this would also need to have `&& closure_scope['facts']['clientnoop']`.

But... the bigger problem seems to be that `closure_scope` is an empty hash when the function is Deferred :(

Reply:
```suggestion
      return 'DUMMY-NOOP-TOKEN' if Puppet.settings['noop']
```
This appears to be fine in Deferred functions.

Reply:
Oh this is because of the Deferred function? oO Didn't realize that. Testing if it works :)

Reply:
Looks like we should technically use a symbol (although if you use a string it gets converted automatically).
Cherrypick/squash in https://github.com/alexjfisher/puppet-gitlab_ci_runner/commit/0de6b5f8b3c16011d9efdf4222ff6e100e95676e to fix tests.